File: 2019/07/rs/src/main.rs
use itertools::Itertools;
use std::cell::RefCell;
use std::collections::VecDeque;
use std::sync::mpsc::{channel, Sender};

type Message = (usize, i32);

pub struct IntCodeComputer {
    memory: RefCell<Vec<i32>>,
    input: RefCell<VecDeque<i32>>,
    output: RefCell<VecDeque<i32>>,
    pointer: usize,
    pub running: bool,
    index: usize,
    sender: Sender<Message>,
}

impl IntCodeComputer {
    pub fn new(memory: &Vec<i32>, input: Vec<i32>, index: usize, sender: Sender<Message>) -> IntCodeComputer {
        IntCodeComputer {
            memory: RefCell::new(memory.clone()),
            input: RefCell::new(VecDeque::from(input)),
            output: RefCell::new(VecDeque::new()),
            pointer: 0,
            running: true,
            index,
            sender,
        }
    }

    pub fn run(&mut self) -> i32 {
        while self.tick() {}
        self.output.borrow_mut().pop_front().unwrap()
    }

    fn get_parameter(&self, offset: i32, mode: i32) -> i32 {
        let value = self.memory.borrow()[self.pointer + offset as usize];
        match mode {
            0 => self.memory.borrow()[value as usize], // POSITION
            1 => value,                                // IMMEDIATE
            _ => {
                panic!("Unrecognized parameter mode '{}'", mode);
            }
        }
    }

    fn add_input(&mut self, value: i32) {
        self.input.borrow_mut().push_back(value);
    }

    fn get_address(&self, offset: i32) -> usize {
        self.memory.borrow()[self.pointer + offset as usize] as usize
    }

    fn tick(&mut self) -> bool {
        if !self.running {
            return false;
        }
        let instruction = self.memory.borrow()[self.pointer];
        let (opcode, p1_mode, p2_mode) = (
            instruction % 100,
            (instruction / 100) % 10,
            (instruction / 1000) % 10,
        );
        match opcode {
            1 => {
                // ADD
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] =
                    self.get_parameter(1, p1_mode) + self.get_parameter(2, p2_mode);
                self.pointer += 4;
            }
            2 => {
                // MUL
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] =
                    self.get_parameter(1, p1_mode) * self.get_parameter(2, p2_mode);
                self.pointer += 4;
            }
            3 => {
                // INPUT
                if let Some(input) = self.input.borrow_mut().pop_front() {
                    let address = self.get_address(1);
                    self.memory.borrow_mut()[address] = input;
                    self.pointer += 2;
                }
            }
            4 => {
                // OUTPUT
                let value = self.get_parameter(1, p1_mode);
                if let Err(_) = self.sender.send((self.index, value)) {
                    panic!("Error sending message.");
                }
                self.output
                    .borrow_mut()
                    .push_back(value);
                self.pointer += 2;
            }
            5 => {
                // JMP_TRUE
                if self.get_parameter(1, p1_mode) != 0 {
                    self.pointer = self.get_parameter(2, p2_mode) as usize;
                } else {
                    self.pointer += 3;
                }
            }
            6 => {
                // JMP_FALSE
                if self.get_parameter(1, p1_mode) == 0 {
                    self.pointer = self.get_parameter(2, p2_mode) as usize;
                } else {
                    self.pointer += 3;
                }
            }
            7 => {
                // LESS_THAN
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] =
                    if self.get_parameter(1, p1_mode) < self.get_parameter(2, p2_mode) {
                        1
                    } else {
                        0
                    };
                self.pointer += 4;
            }
            8 => {
                // EQUALS
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] =
                    if self.get_parameter(1, p1_mode) == self.get_parameter(2, p2_mode) {
                        1
                    } else {
                        0
                    };
                self.pointer += 4;
            }
            99 => {
                // HALT
                self.running = false;
            }
            _ => panic!("Unknown instruction {} {}", self.pointer, opcode),
        }
        self.running
    }
}

fn run_phases_permutation(memory: &Vec<i32>, phases: Vec<i32>) -> i32 {
    let (sender, _receiver) = channel();
    let mut output = 0;
    for phase in &phases {
        output = IntCodeComputer::new(memory, vec![*phase, output], 0, sender.clone()).run();
    }
    output
}

fn run_feedback_phases_permutation(memory: &Vec<i32>, phases: &Vec<i32>) -> i32 {
    let (sender, receiver) = channel();
    let mut amplifiers: Vec<IntCodeComputer> = phases.into_iter()
        .enumerate()
        .map(|(index, phase)| IntCodeComputer::new(memory, vec![*phase], index, sender.clone())).collect();
    amplifiers[0].add_input(0);
    let mut last_output = 0;
    while amplifiers.iter().any(|ammplifier| ammplifier.running) {
        for amplifier in amplifiers.iter_mut() {
            amplifier.tick();
        }
        let mut message = receiver.try_recv();
        while message.is_ok() {
            let (sender_id, value) = message.unwrap();
            amplifiers[(sender_id + 1) % 5].add_input(value);
            if sender_id == 4 {
                last_output = value;
            }
            message = receiver.try_recv();
        }
    };
    last_output
}


fn solve(memory: &Vec<i32>) -> (i32, i32) {
    (
        (0..5)
            .into_iter()
            .permutations(5)
            .map(|phases| run_phases_permutation(memory, phases))
            .max()
            .unwrap(),
        (5..10)
            .into_iter()
            .permutations(5)
            .map(|phases| run_feedback_phases_permutation(memory, &phases))
            .max()
            .unwrap(),
    )
}

fn get_input(file_path: &String) -> Vec<i32> {
    std::fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .split(',')
        .map(|i| i.parse().unwrap())
        .collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:.7}", now.elapsed().as_secs_f32());
}

File: 2019/07/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <string.h>

typedef struct
{
    int *memory, size, capacity;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

#pragma region Queue
typedef struct QueueNode
{
    int value;
    struct QueueNode *next;

} QueueNode;

QueueNode *enqueue(QueueNode *head, int value)
{
    if (head == NULL)
    {
        QueueNode *node = malloc(sizeof(QueueNode));
        node->value = value;
        node->next = NULL;
        head = node;
    }
    else
        head->next = enqueue(head->next, value);
    return head;
}

int isQueueEmpty(QueueNode *head)
{
    return head == NULL;
}

QueueNode *dequeue(QueueNode *head, int *value)
{
    if (head == NULL)
    {
        perror("dequeing from empty queue");
        exit(1);
    }
    else
    {
        *(value) = head->value;
        QueueNode *next = head->next;
        free(head);
        return next;
    }
}

void freeQueue(QueueNode *node)
{
    if (node != NULL)
    {
        freeQueue(node->next);
        free(node);
    }
}
#pragma endregion

#pragma region IntCodeComputer
typedef struct
{
    int *memory;
    int pointer;
    int running;
    QueueNode *input;
    QueueNode *output;
} IntCodeComputer;

IntCodeComputer newComputer(Input memory)
{
    int *newMemory = malloc(memory.size * sizeof(int));
    memcpy(newMemory, memory.memory, memory.size * sizeof(int));
    return (IntCodeComputer){
        newMemory,
        0,
        1,
        NULL,
        NULL};
}

char errorMessage[64];

int getParameter(IntCodeComputer *computer, int offset, int mode)
{
    int value = computer->memory[computer->pointer + offset];
    switch (mode)
    {
    case 0: // POSITION
        return computer->memory[value];
    case 1: // IMMEDIATE
        return value;
    }
    sprintf(errorMessage, "Unrecognized parameter mode '%d'\n", mode);
    perror(errorMessage);
    exit(1);
}

int getAddress(IntCodeComputer *computer, int offset)
{
    return computer->memory[computer->pointer + offset];
}

int tick(IntCodeComputer *computer)
{
    if (!computer->running)
        return 0;
    int instruction = computer->memory[computer->pointer];
    int opcode = instruction % 100, p1Mode = (instruction / 100) % 10, p2Mode = (instruction / 1000) % 10;
    switch (opcode)
    {
    case 1: // ADD
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1, p1Mode) + getParameter(computer, 2, p2Mode);
        computer->pointer += 4;
        break;
    case 2: // MUL
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1, p1Mode) * getParameter(computer, 2, p2Mode);
        computer->pointer += 4;
        break;
    case 3: // INPUT
        if (!isQueueEmpty(computer->input))
        {
            int value;
            computer->input = dequeue(computer->input, &(computer->memory[getAddress(computer, 1)]));
            computer->pointer += 2;
        }
        break;
    case 4: // OUTPUT
        computer->output = enqueue(computer->output, getParameter(computer, 1, p1Mode));
        computer->pointer += 2;
        break;
    case 5: // JMP_TRUE
        if (getParameter(computer, 1, p1Mode))
            computer->pointer = getParameter(computer, 2, p2Mode);
        else
            computer->pointer += 3;
        break;
    case 6: // JMP_FALSE
        if (!getParameter(computer, 1, p1Mode))
            computer->pointer = getParameter(computer, 2, p2Mode);
        else
            computer->pointer += 3;
        break;
    case 7: // LESS_THAN
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1, p1Mode) < getParameter(computer, 2, p2Mode)
                                                        ? 1
                                                        : 0;
        computer->pointer += 4;
        break;
    case 8: // LESS_THAN
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1, p1Mode) == getParameter(computer, 2, p2Mode)
                                                        ? 1
                                                        : 0;
        computer->pointer += 4;
        break;
    case 99: // HALT
        computer->running = 0;
        break;
    default:
        sprintf(errorMessage, "Unknow instruction '%d' at '%d'\n", opcode, computer->pointer);
        perror(errorMessage);
        exit(1);
        break;
    }
    return computer->running;
}

void addToComputerInput(IntCodeComputer *computer, int value)
{
    computer->input = enqueue(computer->input, value);
}

void freeComputer(IntCodeComputer *computer)
{
    free(computer->memory);
    freeQueue(computer->input);
}
#pragma endregion

int run(IntCodeComputer *computer)
{
    while (tick(computer))
        ;
    int result;
    dequeue(computer->output, &result);
    freeComputer(computer);
    return result;
}

int next_permutation(int *indexes, int count)
{
#define swap(i, j)               \
    {                            \
        t = indexes[i];          \
        indexes[i] = indexes[j]; \
        indexes[j] = t;          \
    }
    int k, l, t;
    for (k = count - 1; k && indexes[k - 1] >= indexes[k]; k--)
        ;
    if (!k--)
        return 0;
    for (l = count - 1; indexes[l] <= indexes[k]; l--)
        ;
    swap(k, l);
    for (k++, l = count - 1; l > k; l--, k++)
        swap(k, l);
    return 1;
#undef swap
}

int runPhasesPermutation(Input input, int *phases)
{
    int output = 0;
    for (int phaseIndex = 0; phaseIndex < 5; phaseIndex++)
    {
        IntCodeComputer computer = newComputer(input);
        addToComputerInput(&computer, phases[phaseIndex]);
        addToComputerInput(&computer, output);
        output = run(&computer);
    }
    return output;
}

int getMaxFromPermutations(Input input, int (*runningFunc)(Input, int *))
{
    int phases[5], phaseIndex;
    for (phaseIndex = 0; phaseIndex < 5; phaseIndex++)
        phases[phaseIndex] = phaseIndex;
    int max = 0, result;
    do
    {
        max = (result = runningFunc(input, phases)) > max ? result : max;
    } while (next_permutation(phases, 5));
    return max;
}

int anyRunning(IntCodeComputer *amplifiers, int count)
{
    while (count--)
        if ((amplifiers++)->running)
            return 1;
    return 0;
}

int runFeedbackPhasesPermutation(Input input, int *phases)
{
    IntCodeComputer amplifiers[5];
    int phaseIndex, output, lastOutput;
    for (phaseIndex = 0; phaseIndex < 5; phaseIndex++)
    {
        amplifiers[phaseIndex] = newComputer(input);
        addToComputerInput(&amplifiers[phaseIndex], phases[phaseIndex] + 5);
    }
    addToComputerInput(&amplifiers[0], 0);
    while (anyRunning(amplifiers, 5))
    {
        for (phaseIndex = 0; phaseIndex < 5; phaseIndex++)
        {
            tick(&(amplifiers[phaseIndex]));
            if (!isQueueEmpty(amplifiers[phaseIndex].output))
            {
                amplifiers[phaseIndex].output = dequeue(amplifiers[phaseIndex].output, &output);
                if (phaseIndex == 4)
                    lastOutput = output;
                addToComputerInput(&amplifiers[(phaseIndex + 1) % 5], output);
            }
        }
    }
    for (phaseIndex = 0; phaseIndex < 5; phaseIndex++)
        freeComputer(&(amplifiers[phaseIndex]));
    return lastOutput;
}

Results solve(Input input)
{
    return (Results){
        getMaxFromPermutations(input, runPhasesPermutation),
        getMaxFromPermutations(input, runFeedbackPhasesPermutation)};
}

#define INPUT_INCREMENT 10
void addToInput(Input *input, int value)
{
    if (input->size == input->capacity)
    {
        input->capacity += INPUT_INCREMENT;
        input->memory = realloc(input->memory, input->capacity * sizeof(int));
    }
    input->memory[input->size++] = value;
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *content = malloc(length);
    fread(content, 1, length, file);
    Input input = {
        calloc(INPUT_INCREMENT, sizeof(int)),
        0,
        INPUT_INCREMENT};
    char *value = strtok(content, ",");
    while (value != NULL)
    {
        addToInput(&input, atoi(value));
        value = strtok(NULL, ",");
    }
    fclose(file);
    return input;
}

void freeInput(Input input)
{
    free(input.memory);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
